### ISSUE TITLE
Memory Leak in `VmModule VmModule::FromFlatbufferBlob` using  Python Bindings 

### ISSUE THREAD

**Issue by KoolJBlack** [iree-org/iree#13742]
### What happened?

While tuning 10,000+ kernels using the IREE python bindings, after several hours (+/- days) of running, the process will OOM:

```shell
[Sun May 21 23:59:29 2023] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=/,mems_allowed=0,global_oom,task_memcg=/user.slice/user-1007.slice/session-6.scope,task=python3,pid=445577,uid=1007
[Sun May 21 23:59:29 2023] Out of memory: Killed process 445577 (python3) total-vm:91281012kB, anon-rss:85251192kB, file-rss:74836kB, shmem-rss:8192kB, UID:1007 pgtables:168244kB oom_score_adj:0
```
`tracemalloc` shows persistent growth in memory from `subprocess` beyond invocation of IREE tools, leaking from the use of `compile_str`. 
```shell
/usr/lib/python3.10/subprocess.py:2035: size=8404 KiB, count=11, average=764 KiB
...
  File "/usr/local/google/home/kooljblack/Code/iree-samples/tuning/profiler.py", line 52
    flatbuffer_blob = ireec.compile_str(
  File "/usr/local/google/home/kooljblack/Code/iree-build/compiler/bindings/python/iree/compiler/tools/core.py", line 278
    result = invoke_immediate(cl, immediate_input=input_bytes)
  File "/usr/local/google/home/kooljblack/Code/iree-build/compiler/bindings/python/iree/compiler/tools/binaries.py", line 192
    process = subprocess.run(command_line, capture_output=True, **run_args)
  File "/usr/lib/python3.10/subprocess.py", line 505
    stdout, stderr = process.communicate(input, timeout=timeout)
  File "/usr/lib/python3.10/subprocess.py", line 1154
    stdout, stderr = self._communicate(input, endtime, timeout)
  File "/usr/lib/python3.10/subprocess.py", line 2035
    stdout = b''.join(stdout)
```
### Cause
The runtime bindings rely on the creation of a `VmModule` via pybind11 to expose functions and invocation. `VmModule VmModule::FromFlatbufferBlob()` adds a reference to any passed flatbuffer object: https://github.com/openxla/iree/blob/161daa7f21ea06de46582e24595a224bf24ce611/runtime/bindings/python/vm.cc#L170

In my case, this buffer built on `stdout` from `iree-compile`, but any passed in flatbuffer will leak a reference. Deleting python objects will not release the reference. There is no equivalent function in the bindings to release the reference.

**Comment by benvanik** [comment]
Are you 100% positive about that? the ctl_fn immediately above that is what should be decreasing the reference count.

**Comment by ScottTodd** [comment]
Can you run with ASan/LSan?

**Comment by KoolJBlack** [comment]
> Are you 100% positive about that? the ctl_fn immediately above that is what should be decreasing the reference count.

Its possible I'm using the API incorrectly. Here is a simple case of creating a new buffer and going through the VM to show what happens: 

```python
    flatbuffer_blob_copy = bytearray()
    print(f"ref new buffer: {sys.getrefcount(flatbuffer_blob_copy)}")
    for b in flatbuffer_blob:
        flatbuffer_blob_copy.append(b)
    print(f"ref count after fill: {sys.getrefcount(flatbuffer_blob_copy)}")
    # Create a module
    config = ireert.Config(driver_name=driver)
    vm_module = ireert.VmModule.from_flatbuffer(
        config.vm_instance, flatbuffer_blob_copy)
    print(f"ref count after vm: {sys.getrefcount(flatbuffer_blob_copy)}")
    del config
    del vm_module
    print(f"ref count after del: {sys.getrefcount(flatbuffer_blob_copy)}")
```

and this is the printout: 
```
ref new buffer: 2
ref count after fill: 2
ref count after vm: 4
ref count after del: 3
```

I can't explain why the count starts at 2 instead of1, but if I create a context then module then delete both, its now 3, not 2. Can't show the ref count if I also del `flatbuffer_blob_copy` but I have done this and still get the leak. 

Also, through process of elimination, I'm certain the cause of the leak is passing flatbuffers to the VM module. emitting this code but keeping everything else including compile results in 0 leaks.

**Comment by benvanik** [comment]
Put a printf in the ctl_fn.

**Comment by ScottTodd** [comment]
For ASan, see https://github.com/openxla/iree/issues/9067

**Comment by KoolJBlack** [comment]
> For ASan, see #9067

Looks like running ASAN with the CUDA driver doesn't work. Once I enabled buidling with shared libraries and used the fix from Stellas comment, I ran into this error on VM creation: 
```
Traceback (most recent call last):
  File "/usr/local/google/home/kooljblack/Code/iree-build/runtime/bindings/python/iree/runtime/system_setup.py", line 84, in get_first_device
    return get_device(device_uri, cache=cache)
  File "/usr/local/google/home/kooljblack/Code/iree-build/runtime/bindings/python/iree/runtime/system_setup.py", line 50, in get_device
    device = driver.create_device_by_uri(device_uri)
RuntimeError: Error creating device: iree/runtime/src/iree/hal/drivers/cuda/cuda_driver.c:123: INTERNAL; CUDA driver error 'CUDA_ERROR_OUT_OF_MEMORY' (2): out of memory; creating device 'cuda'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/google/home/kooljblack/Code/iree-samples/tuning/tuner.py", line 130, in <module>
    main(parse_arguments())
  File "/usr/local/google/home/kooljblack/Code/iree-samples/tuning/tuner.py", line 122, in main
    run_program(
  File "/usr/local/google/home/kooljblack/Code/iree-samples/tuning/tuner.py", line 53, in run_program
    run_profile(
  File "/usr/local/google/home/kooljblack/Code/iree-samples/tuning/profiler.py", line 270, in run_profile
    benchmark_results, benchmark_err = run_benchmark_module(compilation_result.flatbuffer_blob, entry_function="forward",
  File "/usr/local/google/home/kooljblack/Code/iree-samples/tuning/profiler.py", line 130, in run_benchmark_module
    config = ireert.Config(driver_name=driver)
  File "/usr/local/google/home/kooljblack/Code/iree-build/runtime/bindings/python/iree/runtime/system_api.py", line 71, in __init__
    self.device = get_first_device(
  File "/usr/local/google/home/kooljblack/Code/iree-build/runtime/bindings/python/iree/runtime/system_setup.py", line 95, in get_first_device
    raise RuntimeError("Could not create device. "
RuntimeError: Could not create device. Exception for last tried follows.
```

To reiterate above, I'm confident the leak is in the VM module since pythons memory analysis `tracemalloc` shows exactly whats leaking and `gc` illustrates where its happening. I'll try ben's print suggestion.

**Comment by benvanik** [comment]
Try with CPU - there's nothing CUDA specific here.

**Comment by KoolJBlack** [comment]
So, building with ASAN did not reveal any memory leaks. I tried this without modifying the code, but then also removing the ctl_fn() and then removing ` flatbuffer_blob.inc_ref();`, still no leaks. I gather that because the leak happens at the python interpreter level, its not something ASAN would track per se? 

Anyway, as expected removing ` flatbuffer_blob.inc_ref()` indeed fixes the leak. The new ref counts from the code in my earlier comment: 
```
ref new buffer: 2
ref count after fill: 2
ref count after vm: 3
ref count after del: 2
```
Count goes up one when bound to the module then resets back to 2 when deleted. Clearly this is expected behavior. 

But without the manual reference increase, what actually ups the count by 1? Turns out assigning the `stashed_flatbuffer_blob`:https://github.com/openxla/iree/blob/main/runtime/bindings/python/vm.cc/#LL184C1-L184C1
pybind is somehow smart enough to automatically handle references on C++ assignment and deletion. There isn't much documentation on how this works however. 

If I remove this, no ref count change in python: 
```
ref new buffer: 2
ref count after fill: 2
ref count after vm: 2
ref count after del: 2
```

@benvanik I checked the ctl_fn, it is getting called. From what I understand, this is not part of pybinds but instead the python's C Api: https://docs.python.org/3/c-api/refcounting.html. From what I gather, the free is happening on the raw PTR to the block data within the bytebuffer object, not the bytbuffer itself. The macro silently does nothing in this case and removing ctl_fn has no change in reference counts. 

Anyway, solution here seems simple. Just remove the superfluous `flatbuffer_blob.inc_ref(); `

**Comment by benvanik** [comment]
That's not superfluous - that's what keeps the memory live for the duration the module hangs on to it.

**Comment by KoolJBlack** [comment]
as it is right now, assigning the flatbuffer object _and_ manually increasing the ref count results in refcount +=2. But deleting the VmModule only results in refcount -=1:
```python
ref count after fill: 2
ref count after vm: 4 # Double ref count from a single VM Call
```

To rephrase my point: Its not clear why we need to double increment this object's reference count when its only held once in the VM? Assigning it to the VmModule will keep it alive for the duration of the VmModule. Also, it won't die during the call to `FromFlatbufferBlob` since the python object also has a live reference in the callees stack frame. 

how are you suggesting this should be counted?

**Comment by benvanik** [comment]
What line 170 (the increment) and line 167 (the decrement) are doing is conceptually correct and must not be removed. If you pass data to iree_vm_bytecode_module_create you must retain that data and then release it in the allocation callback (ctl_fn). Any other retain/release is orthogonal and must not be relied upon for correctness. Printf'ing the reference count at certain points in time and saying "we could probably get away without one of them in some cases" is an extremely yolo approach to working with GC'ed/reference counted systems ;)

IOW: retaining the backing flatbuffer memory for the lifetime of the iree_vm_module_t (_which may outlive the C++ VmModule!_) is critical, that's what the code is supposed to be doing, but maybe it's not and that's what needs to be figured out.

Reading the code I suspect the bug is that it's casting the pointer to be freed to a PyObject* before decrementing the reference - which obviously isn't valid as the pointer to be freed is the flatbuffer data and not a python object. I'm going to guess that instead of passing NULL for self on L171 you'll want to pass `flatbuffer_blob_object.ptr()` and then do `Py_XDECREF(static_cast<PyObject*>(self))` in the ctl_fn and completely ignore the `inout_ptr`. This just looks like a latent 3 year old bug.

**Comment by KoolJBlack** [comment]
Found the reference for how reference counting on PyObjects is handled automatically: https://pybind11.readthedocs.io/en/stable/reference.html#classobject
Reference calls out: 
> Manually increase the reference count of the Python object. Usually, it is preferable to use the [object](https://pybind11.readthedocs.io/en/stable/reference.html#_CPPv46object) class which derives from [handle](https://pybind11.readthedocs.io/en/stable/reference.html#_CPPv46handle) and calls this function automatically. Returns a reference to itself.

This follows what I've found above from experiment. @benvanik  my point is that the reference counting to keep the flatbuffer alive is already happening on assignment and deallocation through pybind. ctl_fn only runs when the object is deallocated, also when pybind releases its reference.

 It not clear why we'd do this separate manually on top of that, which we're currently not clearing properly. rather than try to balance, I'd like to remove the extra reference. Am I missing something?
